userspace/libsinsp/container.cpp

```diff
@@ -139,23 +130,162 @@ size_t sinsp_container_engine_docker::curl_write_callback(const char* ptr, size_
 }
 #endif
 
-bool sinsp_container_engine_docker::parse_docker(sinsp_container_manager* manager, sinsp_container_info *container, sinsp_threadinfo* tinfo)
+void sinsp_container_engine_docker::refresh()
 {
-	string json;
-#ifndef CYGWING_AGENT
-	sinsp_docker_response resp = get_docker(manager, "http://localhost" + m_api_version + "/containers/" + container->m_id + "/json", json);
-#else
-	sinsp_docker_response resp = get_docker(manager, "GET /v1.30/containers/" + container->m_id + "/json HTTP/1.1\r\nHost: docker\r\n\r\n", json);
+#if !defined(CYGWING_AGENT) && defined(HAS_CAPTURE)
+
+	if(m_pending_requests.empty())
+	{
+		return;
+	}
+
+	uint32_t count = 0;
+	while (count < m_n_polling_attempts)
+	{
+		int running_handles;
+		CURLMcode res = curl_multi_perform(m_curlm, &running_handles);
+		if(res != CURLM_OK)
+		{
+			ASSERT(false);
+			return;
+		}
+
+		CURLMsg *m;
+		do
+		{
+			int msgq;
+			m = curl_multi_info_read(m_curlm, &msgq);
+			if(m && (m->msg == CURLMSG_DONE))
+			{
+				CURL *e = m->easy_handle;
+
+				char *url = NULL;
+				if(curl_easy_getinfo(e, CURLINFO_EFFECTIVE_URL, &url) != CURLE_OK)
+				{
+					ASSERT(false);
+					return;
+				}
+
+				if(curl_multi_remove_handle(m_curlm, e) != CURLM_OK)
+				{
+					ASSERT(false);
+					return;
+				}
+
+				docker_request_info *req;
+				if(curl_easy_getinfo(e, CURLINFO_PRIVATE, &req) != CURLE_OK)
+				{
+					ASSERT(false);
+					return;
+				}
+				long http_code = 0;
+				if(curl_easy_getinfo(e, CURLINFO_RESPONSE_CODE, &http_code) != CURLE_OK)
+				{
+					ASSERT(false);
+					return;
+				}
+
+				curl_easy_cleanup(e);
+
+				if(http_code != 200)
+				{
+					if(!m_api_version.empty())
+					{
+						m_api_version.clear();
+						req->url = "http://localhost/containers/" + req->container_info.m_id + "/json";
+						start_docker_request(req);
+					}
+					else
+					{
+						ASSERT(false);
```

Comment:
There's now a possibility of a partial failure where a container request starts and completes, then an image request starts and fails (maybe due to some temporary failure). Will the pending request be properly cleaned up in this case?

Reply:
No it's not, will fix